ArcGISRuntimeSDKQt_CppSamples/Routing/NavigateRoute/NavigateRoute.cpp

```diff
@@ -0,0 +1,271 @@
+// [Legal]
+// Copyright 2020 Esri.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// [Legal]
+
+#ifdef PCH_BUILD
+#include "pch.hpp"
+#endif // PCH_BUILD
+
+#include "NavigateRoute.h"
+
+#include "DefaultLocationDataSource.h"
+#include "DirectionManeuverListModel.h"
+#include "GeometryEngine.h"
+#include "GraphicsOverlay.h"
+#include "Location.h"
+#include "LocationDisplay.h"
+#include "Map.h"
+#include "MapQuickView.h"
+#include "NavigationTypes.h"
+#include "Point.h"
+#include "Route.h"
+#include "RouteResult.h"
+#include "RouteTask.h"
+#include "RouteTracker.h"
+#include "SimpleMarkerSymbol.h"
+#include "SimulatedLocationDataSource.h"
+#include "Stop.h"
+#include "TrackingDistance.h"
+#include "TrackingProgress.h"
+#include "TrackingStatus.h"
+#include "VoiceGuidance.h"
+
+#include <memory>
+#include <QList>
+#include <QGeoPositionInfo>
+#include <QTextToSpeech>
+#include <QTime>
+#include <QUrl>
+
+using namespace Esri::ArcGISRuntime;
+
+namespace  {
+const QUrl routeTaskUrl("https://sampleserver6.arcgisonline.com/arcgis/rest/services/NetworkAnalysis/SanDiego/NAServer/Route");
+const Point conventionCenterPoint(-117.160386727, 32.706608, SpatialReference::wgs84());
+const Point memorialPoint(-117.173034, 32.712327, SpatialReference::wgs84());
+const Point aerospaceMuseumPoint(-117.147230, 32.730467, SpatialReference::wgs84());
+}
+
+NavigateRoute::NavigateRoute(QObject* parent /* = nullptr */):
+  QObject(parent),
+  m_map(new Map(Basemap::navigationVector(this), this))
+{
+  m_routeOverlay = new GraphicsOverlay(this);
+  m_routeTask = new RouteTask(routeTaskUrl, this);
+  m_speaker = new QTextToSpeech(this);
+}
+
+NavigateRoute::~NavigateRoute() = default;
+
+void NavigateRoute::init()
+{
+  // Register the map view for QML
+  qmlRegisterType<MapQuickView>("Esri.Samples", 1, 0, "MapView");
+  qmlRegisterType<NavigateRoute>("Esri.Samples", 1, 0, "NavigateRouteSample");
+}
+
+MapQuickView* NavigateRoute::mapView() const
+{
+  return m_mapView;
+}
+
+// Set the view (created in QML)
+void NavigateRoute::setMapView(MapQuickView* mapView)
+{
+  if (!mapView || mapView == m_mapView)
+    return;
+
+  m_mapView = mapView;
+  m_mapView->setMap(m_map);
+
+  m_mapView->graphicsOverlays()->append(m_routeOverlay);
+  connectRouteTaskSignals();
+
+  m_routeTask->load();
+
+  // add graphics for the predefined stops
+  SimpleMarkerSymbol* stopSymbol = new SimpleMarkerSymbol(SimpleMarkerSymbolStyle::Diamond, Qt::red, 20, this);
```

Comment:
optional - but if you're using the same symbol for many graphics in a graphicsOverlay it is more efficient to use a `Renderer` for the whole overlay